SI-8779 Enable inlining of code within a REPL session

retronym opened the pull request:
The REPL has a long running instance of Global which outputs
classfiles by default to a VirtualDirectory. The inliner did not find
any of these class files when compiling calls to methods defined in
previous runs (ie, previous lines of input.)

This commit:

  - Adds a hook to augment the classpath that the optimizer searches,
    and uses this in the REPL to add the output directory
  - Fixes the implementation of `findClassFile` in VirtualDirectory,
    which doesn't seem to have been used in anger before. I've factored out
    some common code into a new method on `AbstractFile`.
  - Fixes a similar problem getSubDir reported by Li Haoyi
  - Adds missing unit test coverage.

This also fixes a bug in REPL autocompletion for types defined
in packages >= 2 level deep (with the `:paste -raw` command).
I've added a test for this case.

Review by @lrytz

retronym committed:
SI-8779 Enable inlining of code within a REPL session

The REPL has a long running instance of Global which outputs
classfiles by default to a VirtualDirectory. The inliner did not find
any of these class files when compiling calls to methods defined in
previous runs (ie, previous lines of input.)

This commit:

  - Adds a hook to augment the classpath that the optimizer searches,
    and uses this in the REPL to add the output directory
  - Fixes the implementation of `findClassFile` in VirtualDirectory,
    which doesn't seem to have been used in anger before. I've factored out
    some common code into a new method on `AbstractFile`.
  - Fixes a similar problem getSubDir reported by Li Haoyi
  - Adds missing unit test coverage.

This also fixes a bug in REPL autocompletion for types defined
in packages >= 2 level deep (with the `:paste -raw` command).
I've added a test for this case.

som-snytt commented:
The sun rises early down under.

Missed my shot at this; and Li H. just complained about findClassFile; the lesson is, you can't wait for the weekend.

I once had a repl that specialized the platform classpath, with the old classpath impl; I guess putting the hook on global is much simpler.

som-snytt commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -6,12 +6,16 @@
  *  functionality for the repl.  It doesn't do much yet.
  */
 trait ReplGlobal extends Global {
+  self =>
```
To thine own `self =>` be true.

lihaoyi commented:
FWIW, I'm pretty sure `getSubDir` on the `VirtualDirectoryClassPath` is borked too. It suffers from the same problem: doing a shallow lookup of the immediate children, rather than recursively resolving each path segment to find the file it wants.

This is what I had to do to get 2.12's `VirtualDirectoryClassPath` working with Ammonite:

```scala
    val dynamicCP = new VirtualDirectoryClassPath(dynamicClasspath){

      override def getSubDir(packageDirName: String): Option[AbstractFile] = {
        val pathParts = packageDirName.split('/')
        var file: AbstractFile = dir
        for (dirPart <- pathParts) {
          file = file.lookupName(dirPart, directory = true)
          if (file == null) return None
        }
        Some(file)

      }
      override def findClassFile(className: String): Option[AbstractFile] = {
        val relativePath = FileUtils.dirPath(className)
        val pathParts = relativePath.split('/')
        var file: AbstractFile = dir
        for (dirPart <- pathParts.init) {
          file = file.lookupName(dirPart, directory = true)
          if (file == null) return None
        }

        file.lookupName(pathParts.last + ".class", directory = false) match {
          case null => None
          case file => Some(file)
        }
      }

    }
```

There may be other issues with it, but all my tests pass, so that's good enough for me for now.

(FWIW, `findClassFile` doesn't seem to be picked up in the normal flow of compilation, but i included the fix anyway because the original code was clearly incorrect and maybe some future version will start using it)

retronym commented:
@lihaoyi Thanks for the heads up. I've fixed that method as well. Autocompletion in the REPL was actually broken because of this for classes defined with `:paste` in nested packages.

som-snytt commented:
With an existing `Thing` on the class path:
```
$ ./build/pack/bin/scala -cp /tmp
Welcome to Scala 2.12.1-20161121-143919-1d2d383 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_111).
Type in expressions for evaluation. Or try :help.

scala> stuff.Thing
   object Thing

scala> stuff.Thing.f
res0: Int = 42

scala> :pa
// Entering paste mode (ctrl-D to finish)

package stuff { object Thing { def g = "hi" } }

// Exiting paste mode, now interpreting.


scala> stuff.Thing.g    // autocompletes!
java.lang.NoSuchMethodError: stuff.Thing$.g()Ljava/lang/String;
  ... 31 elided


```

som-snytt commented on src/compiler/scala/tools/nsc/backend/JavaPlatform.scala:
```diff
@@ -23,6 +23,7 @@ trait JavaPlatform extends Platform {
     if (currentClassPath.isEmpty) currentClassPath = Some(new PathResolver(settings).result)
     currentClassPath.get
   }
+  private[nsc] def previousRunClassPath: Option[ClassPath]= None
```
?

retronym replied:
removed.

som-snytt commented on src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala:
```diff
@@ -290,7 +290,9 @@ class ByteCodeRepository[BT <: BTypes](val classPath: ClassPath, val btypes: BT)
-      case None       => Left(ClassNotFound(internalName, javaDefinedClasses(internalName)))
+      case None       =>
+
+        Left(ClassNotFound(internalName, javaDefinedClasses(internalName)))
```
`Left(some extra space)`.

retronym replied:
:)

`Right(back to where we started)`

som-snytt commented:
I hadn't noticed the nested package quirk.
```
scala> :pa
// Entering paste mode (ctrl-D to finish)

package p { package q { object x { def f = 42 } } }

// Exiting paste mode, now interpreting.


scala> p.q.x
   object x

scala> p.q.x.f   // completed
res1: Int = 42

```

som-snytt commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -6,6 +6,9 @@
 package scala.tools.nsc
 package interpreter
 
+import scala.tools.nsc.backend.JavaPlatform
```
This is the one you obviate.

som-snytt commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -31,4 +34,14 @@ trait ReplGlobal extends Global {
+      case Some(out) =>
+        // Make bytecode of previous lines available to the inliner
+        val replOutClasspath = ClassPathFactory.newClassPath(settings.outputDirs.getSingleOutput.get, settings)
+        AggregateClassPath(platform.classPath :: replOutClasspath :: Nil)
```
This order made me think twice. The REPL class loader delegates to parent; it would be a feature to prefer REPL output to whatever is on the class path. But it's a little late in the day (local time) to start pondering that.

retronym replied:
A more conservative feature might be to warn that the generated classfiles from `:paste package ...` clash with existing classpath entries and won't be picked up.

som-snytt commented:
LGTM

retronym commented:
I'll let @lrytz take a look at this from the inliner perspective before merging.

som-snytt commented:
This isn't surprising, but I noticed that module loads are not optimized away; I think it was considered important to keep side-effects of module init for case class apply rewriting.

And I notice that `-Yrepl-class-based` class wrappers are not final, as it will warn for inlining a non-final method.

retronym committed:
Support inlining under -Yrepl-class-based REPL

By marking the wrapper classes as sealed, the inliner will be
able to assume finality of defs introduces in the REPL without
requiring the user to mark them as `final`, which is an odd thing
to do in single line of REPL input.

retronym commented:
> And I notice that -Yrepl-class-based class wrappers are not final, as it will warn for inlining a non-final method.

Just pushed a commit to make this work, by marking them `sealed`.

retronym commented:
> This isn't surprising, but I noticed that module loads are not optimized away; I think it was considered important to keep side-effects of module init for case class apply rewriting.

Yeah, in the REPL context we need to take another swing at getting rid of those pesky wrappers altogether. As it stands, the REPL gets a little slower every line by needing ever deeper wrappers.

som-snytt commented:
@retronym My PR for your unwrapped REPL is dormant but I'll revive it. After a PR is several months old, it's real work just to take it up again.

lrytz commented:
needs a check file update for `test/files/run/t7747-repl.check` - i tried pushing to your branch, but permission denied

lrytz commented on src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala:
```diff
@@ -31,4 +34,14 @@ trait ReplGlobal extends Global {
+      case Some(out) =>
+        // Make bytecode of previous lines available to the inliner
+        val replOutClasspath = ClassPathFactory.newClassPath(settings.outputDirs.getSingleOutput.get, settings)
+        AggregateClassPath(platform.classPath :: replOutClasspath :: Nil)
```
Could use `AggregateClassPath.createAggregate` on this line to prevent nesting `AggregateClassPath`s

lrytz commented on test/files/run/repl-inline.scala:
```diff
@@ -0,0 +1,27 @@
+      val settings = new Settings()
+      settings.processArgumentString("-opt:l:classpath")
+      f(settings)
+      settings.classpath.value = System.getProperty("java.class.path")
```
could we instead pass `"-usejavacp"` above?

lrytz commented on test/junit/scala/reflect/internal/util/AbstractFileClassLoaderTest.scala:
```diff
@@ -5,6 +5,9 @@ import org.junit.Test
 
+import scala.tools.nsc.classpath.VirtualDirectoryClassPath
+import scala.tools.nsc.util.ClassPath
+
```
no-op change

lrytz commented:
looks all good aside from these details.
